### ISSUE TITLE
Compile failure with MSVC and /permissive- (Two-phase lookup)

### ISSUE THREAD

**Issue by kylemacfarlan** [pybind/pybind11#1107]
Pybind11 fails to compile with newer MSVC compilers, under /permissive- mode (extra standards conformance). Disabling the new two-phase lookup (with /Zc:twoPhase-) fixes the issue, but it ought to be able to compile without it. 

`#include <pybind11/pybind11.h>
using namespace pybind11;

double Mult(double a, double b) { return a*b; }

void export_Mult(pybind11::module& m) {
      m.def("mult", &Mult);
}`

produces:
`pybind11\cast.h(1416): error C2187: syntax error: ')' was unexpected here`

**Comment by patrikhuber** [comment]
I am encountering the same. The line that the compiler complains about is this:

```
    bool load_value(value_and_holder &&v_h) {
        if (v_h.holder_constructed()) {
            value = v_h.value_ptr();
            holder = v_h.holder<holder_type>(); // <-- syntax error: ')' was unexpected here
            return true;
        } else {
            throw cast_error("Unable to cast from non-held to held instance (T& to Holder<T>) "
#if defined(NDEBUG)
                             "(compile in debug mode for type information)");
#else
                             "of type '" + type_id<holder_type>() + "''");
#endif
        }
    }
```

**Comment by jagerman** [comment]
Try changing it to:
```C++
holder = v_h.template holder<holder_type>();
```
If that fixes it, it's an easy commit.

**Comment by patrikhuber** [comment]
Awesome @jagerman, this fixes the issue, I've created a PR. Hope it works on the other platforms as well.

**Comment by patrikhuber** [comment]
CI is happy, and there's no further errors when compiling with `/permissive-`. Awesome. Thanks! I'd say this can be closed :-)

**Comment by jagerman** [comment]
I'll merge #1155 to fix this once the ci builds finish -- I've updated the VS2017 builds to include the `-permissive-` flag.

Incidentally, I remember reading somewhere that exactly this issue is why the standards committee went for `std::get<N>(tuple)` rather than `tuple.get<N>()` — they didn't want people to have to write `tuple.template get<N>()` which is needed in some contexts (like this one) to avoid ambiguity.  I'd be tempted to follow that pattern here if we needed to fix it in a bunch of places, but just fixing the one line here seems enough.
